pkg/operator/ceph/object/health.go

```diff
@@ -0,0 +1,269 @@
+}
+
+// checkCephStatus periodically checks the health of the cluster
+func (c *rgwStatusChecker) checkObjectStore(stopCh chan struct{}) {
+
+	for {
+		select {
+		case <-stopCh:
+			// purge bucket and s3 user
+			// Needed for external mode where in converged everything goes away with the CR deletion
+			c.cleanupHealthCheck()
+			logger.Infof("stopping monitoring of rgw endpoints")
+			return
+
+		case <-time.After(defaultStatusCheckInterval):
+			logger.Debug("checking rgw health")
+			err := c.checkObjectStoreHealth()
+			if err != nil {
+				logger.Warningf("failed to check rgw health. %v", err)
+			}
+		}
+	}
+}
+
+func (c *rgwStatusChecker) checkObjectStoreHealth() error {
+	/*
+		0. purge everything by default
+		1. create an S3 user
+		2. always use the same user
+		3. if already exists just re-hydrate the s3 credentials
+		4. create a bucket with that user or use the existing one (always use the same bucket)
+		5. create a check file
+		6. get the MD5 of the file
+		7. PUT the file
+		8. GET the file
+		9. compare MD5
+		10. delete object on bucket
+		11. update CR health status check
+
+		Always keep the bucket and the user for the healthcheck, just do PUT and GET because bucket creation is expensive
+	*/
+
+	var s3AccessKey string
+	var s3SecretKey string
+	s3endpoint := fmt.Sprintf("%s:%s", BuildDomainName(c.objContext.Name, c.namespacedName.Namespace), c.port)
+
+	// Generate unique user and bucket name
+	bucket := genUniqueBucketName(c.objContext.UID)
+	userConfig := c.genUserConfig()
+
+	// Create S3 user
+	logger.Debugf("creating s3 user object %q", userConfig.UserID)
+	user, rgwerr, err := CreateUser(c.objContext, userConfig)
+	if err != nil {
+		if rgwerr == ErrorCodeFileExists {
+			objectUser, _, err := GetUser(c.objContext, userConfig.UserID)
+			if err != nil {
+				err = errors.Wrapf(err, "failed to get details from ceph object user %q", objectUser.UserID)
+				updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+				return err
+			}
+
+			// Set access and secret key
+			s3AccessKey = *objectUser.AccessKey
+			s3SecretKey = *objectUser.SecretKey
+
+		} else {
+			err = errors.Wrapf(err, "failed to create s3 user %q. error code %d", userConfig.UserID, rgwerr)
+			updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+			return err
+		}
+	} else {
+		// Set access and secret key
+		s3AccessKey = *user.AccessKey
+		s3SecretKey = *user.SecretKey
+	}
+
+	// Initiate s3 agent
+	logger.Debug("initializing s3 connection")
+	s3client := CreateNewS3Client(s3endpoint, s3AccessKey, s3SecretKey)
+
+	// Force purge before starting anything
+	err = cleanupObjectHealthCheck(s3client, c.objContext.UID)
+	if err != nil {
+		err = errors.Wrap(err, "failed to perform object cleanup")
+		updateStatusEndpoint(c.client, c.namespacedName, cephv1.ConditionFailure, err.Error())
+		return err
+	}
+
+	// Purge on exit
+	defer cleanupObjectHealthCheck(s3client, c.objContext.UID)
```

Comment:
How about creating a new method such as `testBucketHealth()` that puts, gets, and validates? Then the defer will only apply to what happens in that method, and this method won't be so long.